### ISSUE TITLE
[jvm-packages] xgboost4j-spark training failed when running at foreachPartition at XGBoost.scala:287

### ISSUE THREAD

**Issue by CoderLee2014** [dmlc/xgboost#4058]
I'm using spark 2.3.1 and xgboost4j-0.81.
The datasize is about 100G, more than 3000 columns in training dataframe.
Training params for spark:
driver-mem=8g,
executor-num=35,
executor-core=2,
executor-mem=26g,

for xgboostClassifier:
        "eta" -> 0.1f,
        "max_depth" -> 6,
        "objective" -> "binary:logistic",
        "num_round" -> 300,
        "num_workers" -> 30,
        "lambda" -> 10,
        "min_child_weight" -> 3,
        "gamma" -> 1,
        "subsample" -> 0.8,
        "colsample_bytree" -> 0.8
@CodingCat

**Comment by CoderLee2014** [comment]
![image](https://user-images.githubusercontent.com/10073176/51107172-fcc46a80-1828-11e9-9617-0ca3d0d1693d.png)
For failed stages exectuors stdout:
![image](https://user-images.githubusercontent.com/10073176/51107224-1e255680-1829-11e9-9501-d0fbbd98b212.png)

**Comment by CoderLee2014** [comment]
Sometimes, it was stucked at foreachpartition at XGBoost.scala：287
Here is the log:
Tracker started, with env={DMLC_NUM_SERVER=0, DMLC_TRACKER_URI=10.19.70.3, DMLC_TRACKER_PORT=9091, DMLC_NUM_WORKER=30}

in stderr log:
[heartbeat-receiver-event-loop-thread] WARN org.apache.spark.scheduler.AsyncEventQueue - Dropped 208 events from shared since Mon Jan 14 18:15:24 CST 2019.
[heartbeat-receiver-event-loop-thread] WARN org.apache.spark.scheduler.AsyncEventQueue - Dropped 204 events from shared since Mon Jan 14 18:16:24 CST 2019.
[heartbeat-receiver-event-loop-thread] WARN org.apache.spark.scheduler.AsyncEventQueue - Dropped 204 events from shared since Mon Jan 14 18:17:24 CST 2019.
[heartbeat-receiver-event-loop-thread] WARN org.apache.spark.scheduler.AsyncEventQueue - Dropped 204 events from shared since Mon Jan 14 18:18:24 CST 2019.
[heartbeat-receiver-event-loop-thread] WARN org.apache.spark.scheduler.AsyncEventQueue - Dropped 204 events from shared since Mon Jan 14 18:19:24 CST 2019.
[heartbeat-receiver-event-loop-thread] WARN org.apache.spark.scheduler.AsyncEventQueue - Dropped 205 events from shared since Mon Jan 14 18:20:24 CST 2019.
[heartbeat-receiver-event-loop-thread] WARN org.apache.spark.scheduler.AsyncEventQueue - Dropped 204 events from shared since Mon Jan 14 18:21:24 CST 2019.
[heartbeat-receiver-event-loop-thread] WARN org.apache.spark.scheduler.AsyncEventQueue - Dropped 204 events from shared since Mon Jan 14 18:22:24 CST 2019.
[heartbeat-receiver-event-loop-thread] WARN org.apache.spark.scheduler.AsyncEventQueue - Dropped 204 events from shared since Mon Jan 14 18:23:24 CST 2019.
[heartbeat-receiver-event-loop-thread] WARN org.apache.spark.scheduler.AsyncEventQueue - Dropped 205 events from shared since Mon Jan 14 18:24:24 CST 2019.
[heartbeat-receiver-event-loop-thread] WARN org.apache.spark.scheduler.AsyncEventQueue - Dropped 204 events from shared since Mon Jan 14 18:25:24 CST 2019.
[heartbeat-receiver-event-loop-thread] WARN org.apache.spark.scheduler.AsyncEventQueue - Dropped 204 events from shared since Mon Jan 14 18:26:24 CST 2019.
[heartbeat-receiver-event-loop-thread] WARN org.apache.spark.scheduler.AsyncEventQueue - Dropped 204 events from shared since Mon Jan 14 18:27:24 CST 2019.
[heartbeat-receiver-event-loop-thread] WARN org.apache.spark.scheduler.AsyncEventQueue - Dropped 204 events from shared since Mon Jan 14 18:28:24 CST 2019.
[heartbeat-receiver-event-loop-thread] WARN org.apache.spark.scheduler.AsyncEventQueue - Dropped 204 events from shared since Mon Jan 14 18:29:24 CST 2019.
[heartbeat-receiver-event-loop-thread] WARN org.apache.spark.scheduler.AsyncEventQueue - Dropped 204 events from shared since Mon Jan 14 18:30:24 CST 2019.

**Comment by xiejun901** [comment]
Maybe some columns of  you feature vector  are all missing value after subsample, you can try to convert the feature to dense vector or find the columns with many missing value and remove it.

**Comment by CodingCat** [comment]
https://github.com/dmlc/xgboost/pull/4045 fixed the issue
